File: backend/app/parser.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import PurePosixPath
import re
import urllib.parse

VIDEO_EXTENSIONS = {".mkv", ".mp4", ".m4v", ".avi", ".mov", ".ts", ".m2ts", ".webm"}

SEASON_DIR_PATTERNS = [
    re.compile(r"^(?:season|сезон)[ ._-]*(\d{1,2})$", re.I),
    re.compile(r"^s(\d{1,2})$", re.I),
]
EPISODE_PATTERNS = [
    re.compile(r"(?i)(?:^|[ ._\-])S(\d{1,2})E(\d{1,3})(?:E\d{1,3})?"),
    re.compile(r"(?i)(?:^|[ ._\-])(\d{1,2})x(\d{1,3})(?:[ ._\-]|$)"),
]
YEAR_RE = re.compile(r"(?<!\d)((?:19|20)\d{2})(?!\d)")
QUALITY_TOKENS = re.compile(
    r"(?i)\b(?:2160p|1080p|720p|480p|4k|uhd|hdr10\+?|hdr|dv|dolby[ ._-]?vision|"
    r"bluray|blu[ ._-]?ray|bdrip|webrip|web[ ._-]?dl|remux|h\.?26[45]|hevc|x26[45]|"
    r"aac|ac3|eac3|dts(?:-hd)?|truehd|atmos|proper|repack|extended|multi)\b.*$"
)
BRACKET_RE = re.compile(r"\[[^\]]*\]|\([^)]*(?:rip|codec|audio|video|gb|mb)[^)]*\)", re.I)


@dataclass(frozen=True)
class ParsedMedia:
    kind: str  # movie | episode
    title: str
    year: int | None = None
    show_title: str | None = None
    season: int | None = None
    episode: int | None = None


def decode_name(value: str) -> str:
    return urllib.parse.unquote(value).replace("+", " ")


def clean_title(raw: str) -> str:
    value = decode_name(raw)
    value = PurePosixPath(value).stem
    value = BRACKET_RE.sub(" ", value)
    value = QUALITY_TOKENS.sub("", value)
    value = re.sub(r"[._]+", " ", value)
    value = re.sub(r"\s*[-–—]\s*", " ", value)
    value = re.sub(r"\s+", " ", value).strip(" ._-—")
    return value.strip()


def season_from_dir(name: str) -> int | None:
    clean = decode_name(name).strip("/ ")
    for pattern in SEASON_DIR_PATTERNS:
        match = pattern.match(clean)
        if match:
            return int(match.group(1))
    return None


def _episode_from_filename(filename: str) -> tuple[int, int] | None:
    stem = decode_name(PurePosixPath(filename).stem)
    for pattern in EPISODE_PATTERNS:
        match = pattern.search(stem)
        if match:
            return int(match.group(1)), int(match.group(2))
    return None


def _numeric_episode(filename: str) -> int | None:
    stem = decode_name(PurePosixPath(filename).stem).strip()
    match = re.match(r"^(?:e|ep|episode|серия)?[ ._-]*(\d{1,3})(?:\D.*)?$", stem, re.I)
    if match:
        return int(match.group(1))
    return None


def _strip_episode_tokens(title: str) -> str:
    value = title
    for pattern in EPISODE_PATTERNS:
        value = pattern.sub(" ", value)
    value = re.sub(r"(?i)\b(?:episode|ep|серия)\s*\d{1,3}\b", " ", value)
    value = re.sub(r"\s+", " ", value).strip()
    return value


def parse_media_url(url: str, base_url: str) -> ParsedMedia | None:
    parsed = urllib.parse.urlsplit(url)
    path = PurePosixPath(parsed.path)
    if path.suffix.lower() not in VIDEO_EXTENSIONS:
        return None

    base_path = PurePosixPath(urllib.parse.urlsplit(base_url).path or "/")
    try:
        rel = path.relative_to(base_path)
    except ValueError:
        rel = path

    parts = [decode_name(p) for p in rel.parts if p not in ("/", "")]
    if not parts:
        return None

    filename = parts[-1]
    parent_dirs = parts[:-1]

    direct_episode = _episode_from_filename(filename)
    season_dir_index = None
    season_dir_number = None
    for idx in range(len(parent_dirs) - 1, -1, -1):
        number = season_from_dir(parent_dirs[idx])
        if number is not None:
            season_dir_index = idx
            season_dir_number = number
            break

    if direct_episode:
        season, episode = direct_episode
        if season_dir_index is not None and season_dir_index > 0:
            show_title = clean_title(parent_dirs[season_dir_index - 1])
        elif parent_dirs:
            show_title = clean_title(parent_dirs[-1])
        else:
            raw = _strip_episode_tokens(clean_title(filename))
            show_title = raw or "Неизвестный сериал"
        return ParsedMedia(
            kind="episode",
            title=f"Серия {episode}",
            show_title=show_title,
            season=season,
            episode=episode,
        )

    numeric_episode = _numeric_episode(filename)
    if season_dir_index is not None and numeric_episode is not None:
        show_title = (
            clean_title(parent_dirs[season_dir_index - 1])
            if season_dir_index > 0
            else "Неизвестный сериал"
        )
        return ParsedMedia(
            kind="episode",
            title=f"Серия {numeric_episode}",
            show_title=show_title,
            season=season_dir_number,
            episode=numeric_episode,
        )

    raw = clean_title(filename)
    year_match = YEAR_RE.search(raw)
    year = int(year_match.group(1)) if year_match else None
    title = YEAR_RE.sub(" ", raw)
    title = re.sub(r"\s+", " ", title).strip()
    return ParsedMedia(kind="movie", title=title or raw, year=year)

File: native-qnap-d1/internal/app/parser.go
package app

import (
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

var videoExt = map[string]bool{".mkv": true, ".mp4": true, ".m4v": true, ".avi": true, ".mov": true, ".ts": true, ".m2ts": true, ".webm": true}
var epS = regexp.MustCompile(`(?i)(?:^|[ ._\-])s([0-9]{1,2})[ ._\-]*e([0-9]{1,3})(?:[ ._\-]|$)`)
var epX = regexp.MustCompile(`(?i)(?:^|[ ._\-])([0-9]{1,2})x([0-9]{1,3})(?:[ ._\-]|$)`)
var seasonDir = regexp.MustCompile(`(?i)^(?:season|сезон)[ ._\-]*([0-9]{1,2})$`)
var seasonS = regexp.MustCompile(`(?i)^s([0-9]{1,2})$`)
var numericEp = regexp.MustCompile(`(?i)^(?:e|ep|episode|серия)?[ ._\-]*([0-9]{1,3})(?:\D.*)?$`)
var yearRe = regexp.MustCompile(`(?:19|20)[0-9]{2}`)
var qualityRe = regexp.MustCompile(`(?i)\b(?:2160p|1080p|720p|480p|4k|uhd|hdr10\+?|hdr|dv|dolby[ ._\-]?vision|bluray|blu[ ._\-]?ray|bdrip|webrip|web[ ._\-]?dl|dcprip|dcp|remux|amzn|netflix|nf|dsnp|hmax|atvp|h\.?26[45]|hevc|x26[45]|aac|ac3|eac3|dts(?:-hd)?|truehd|atmos|proper|repack|extended)\b.*$`)
var seasonSuffix = regexp.MustCompile(`(?i)(?:^|[ ._\-])(?:s[0-9]{1,2}|season[ ._\-]*[0-9]{1,2}|сезон[ ._\-]*[0-9]{1,2})\b.*$`)
var trailingReleaseTag = regexp.MustCompile(`(?i)(?:[ ._\-]+)(?:ma)(?:[ ._\-]*)$`)

type Parsed struct {
	Kind, Title, ShowTitle string
	Year, Season, Episode  int
}

func seasonNo(s string) int {
	for _, r := range []*regexp.Regexp{seasonDir, seasonS} {
		m := r.FindStringSubmatch(s)
		if len(m) > 1 {
			n, _ := strconv.Atoi(m[1])
			return n
		}
	}
	return -1
}
func cleanTitle(s string) string {
	if ext := strings.ToLower(filepath.Ext(s)); videoExt[ext] {
		s = strings.TrimSuffix(s, filepath.Ext(s))
	}
	s = qualityRe.ReplaceAllString(s, "")
	for trailingReleaseTag.MatchString(s) {
		s = trailingReleaseTag.ReplaceAllString(s, "")
	}
	s = strings.NewReplacer(".", " ", "_", " ", "–", " ", "—", " ").Replace(s)
	s = strings.Join(strings.Fields(s), " ")
	return strings.Trim(s, " ._-")
}
func cleanShowTitle(s string) string {
	s = seasonSuffix.ReplaceAllString(s, "")
	s = cleanTitle(s)
	s = strings.TrimSpace(yearRe.ReplaceAllString(s, ""))
	s = strings.Join(strings.Fields(s), " ")
	return strings.Trim(s, " ._-")
}
func showTitleFromEpisode(stem string, dirs []string, sdIdx int, re *regexp.Regexp) string {
	if loc := re.FindStringIndex(stem); len(loc) == 2 && loc[0] > 0 {
		if title := cleanShowTitle(stem[:loc[0]]); title != "" {
			return title
		}
	}
	if sdIdx > 0 {
		if title := cleanShowTitle(dirs[sdIdx-1]); title != "" {
			return title
		}
	}
	if len(dirs) > 0 {
		if title := cleanShowTitle(dirs[len(dirs)-1]); title != "" {
			return title
		}
	}
	return "Неизвестный сериал"
}
func episodeParsed(show string, season, episode int) Parsed {
	kind := "episode"
	title := "Серия " + strconv.Itoa(episode)
	if localKind, localTitle, ok := localEpisodeOverride(show, season, episode); ok {
		kind = localKind
		title = localTitle
	}
	return Parsed{Kind: kind, Title: title, ShowTitle: show, Season: season, Episode: episode}
}
func ParseMedia(rel string) (Parsed, bool) {
	ext := strings.ToLower(filepath.Ext(rel))
	if !videoExt[ext] {
		return Parsed{}, false
	}
	rel = filepath.ToSlash(rel)
	parts := strings.Split(rel, "/")
	file := parts[len(parts)-1]
	dirs := parts[:len(parts)-1]
	stem := strings.TrimSuffix(file, filepath.Ext(file))
	var sm []string
	var episodeRe *regexp.Regexp
	if x := epS.FindStringSubmatch(stem); len(x) > 0 {
		sm = x
		episodeRe = epS
	} else if x := epX.FindStringSubmatch(stem); len(x) > 0 {
		sm = x
		episodeRe = epX
	}
	sdIdx, sd := -1, -1
	for i := len(dirs) - 1; i >= 0; i-- {
		if n := seasonNo(dirs[i]); n >= 0 {
			sdIdx = i
			sd = n
			break
		}
	}
	if len(sm) > 2 {
		s, _ := strconv.Atoi(sm[1])
		e, _ := strconv.Atoi(sm[2])
		show := showTitleFromEpisode(stem, dirs, sdIdx, episodeRe)
		return episodeParsed(show, s, e), true
	}
	if sdIdx >= 0 {
		if m := numericEp.FindStringSubmatch(stem); len(m) > 1 {
			e, _ := strconv.Atoi(m[1])
			show := "Неизвестный сериал"
			if sdIdx > 0 {
				if title := cleanShowTitle(dirs[sdIdx-1]); title != "" {
					show = title
				}
			}
			return episodeParsed(show, sd, e), true
		}
	}
	title := cleanTitle(file)
	year := 0
	if y := yearRe.FindString(title); y != "" {
		year, _ = strconv.Atoi(y)
		title = strings.TrimSpace(yearRe.ReplaceAllString(title, ""))
	}
	return Parsed{Kind: "movie", Title: title, Year: year}, true
}
